backend/data_preparation/extractor/fire_extractor.py

```diff
@@ -0,0 +1,175 @@
+"""
+@author: Scarlett Zhang
+This file contains 2 classes:
+1. class IncompleteShapefileError: an Exception to be caught in pipeline main function.
+2. class FireExtractor: the object of extractor from shapefile to dictionary with useful information
+"""
+import rootpath
+rootpath.append()
+import re
+from backend.data_preparation.extractor.extractorbase import ExtractorBase
+import shapefile
+import logging
+import datetime
+from typing import Dict, List
+from shapely.geometry import shape
+from shapely.geometry.multipolygon import MultiPolygon
+
+logger = logging.getLogger('TaskManager')
+
+
+class IncompleteShapefileError(Exception):
+    pass
+
+
+class FireExtractor(ExtractorBase):
+
+    def extract(self, path: str, record: str, if_sequence: bool, id: int, state: str) -> Dict[str, str]:
+        """
+        Reads a set of records and return the contents.
+        This is a messy function since the data source is dirty.
+        :param path: str, path of the folder
+        :param record: str, name of the record
+        :param if_sequence: bool, if this fire is a sequence of fire
+        :param id: int
+        :param state: str
+        :return: dict of all information in the record
+        """
+        try:
+            # get the year of this fire record
+            year = int(re.search(r"\d{8}",record).group()[:4])
+        except AttributeError:
+            # there is a wrong formatted timestamp
+            # sorry the data source is messy
+            year = int(re.search(r"\d{7}",record).group()[:4])
+        # defining fields each year in dict is not proper since
+        # before 2016 there are 4 fields needed. But after 2016 there are only 3
+        # decided to use if statement
+        # NOTE: current year's schema is different from 2016-2018, not sure if it is a temp field names for
+        # current_year only or the names of fields starts to change again after 2018
+        # result to return -- a dict
+        result = dict()
+        # read the shapefile
+        try:
+            shp = shapefile.Reader(path + "/" + record)
+        except shapefile.ShapefileException:
+            # if the sub-files of the shapefile is not complete
+            # then it is not a valid shapefile, and no result should be returned
+            raise IncompleteShapefileError
+            # return result
+        # fill result dict based on the format for this year
+        if year < 2016:
+            # FIRE_NAME, DATE_:  datetime.date(2014, 9, 11), TIME_: 0129, AGENCY: USFS or NULL
+            result["firename"] = shp.record(0)["FIRE_NAME"].capitalize()
+            # before 2016, record schema for firename is: FIRE_NAME
+            result["agency"] = shp.record(0)["AGENCY"] if shp.record(0)["AGENCY"] != "" else "Unknown"
+            # before 2016, record schema for agency is: AGENCY
+            try:
+                result["datetime"] = datetime.datetime.strptime("{:%m%d%Y}".format(shp.record(0)["DATE_"]) + shp.record\
```

Comment:
"{:%m%d%Y}".format(shp.record(0)["DATE_"] can be extracted as a variable. and I believe this should be outside try block?